### PR TITLE
account/abi: handle solidity panic revert

### PR DESCRIPTION
**Pull request by islishude**
ref https://docs.soliditylang.org/en/v0.8.21/control-structures.html#panic-via-assert-and-error-via-require

### CODE REVIEW

**Review by holiman — APPROVED** [review]
Thanks, looks good to me!

File: accounts/abi/abi.go [hunk 1]
```diff
@@ -246,24 +247,65 @@ func (abi *ABI) HasReceive() bool {
...
+		}
+		pCode := unpacked[0].(*big.Int)
+		// uint64 safety check for future
+		// but the code is not bigger than MAX(uint64) now
+		if pCode.IsUint64() {
+			if reason, ok := panicReasons[pCode.Uint64()]; ok {
+				return reason, nil
+			}
+		}
+		return fmt.Sprintf("unknown panic code: 0x%s", pCode.Text(16)), nil
```
Comment (rjl493456442) [inline accounts/abi/abi.go:307]: return fmt.Sprintf("unknown panic code: %#x", pCode), nil

File: accounts/abi/abi.go [hunk 2]
```diff
@@ -246,24 +247,65 @@ func (abi *ABI) HasReceive() bool {
...
+	0x41: "out of memory",
+	0x51: "uninitialized function",
+}
+
 // UnpackRevert resolves the abi-encoded revert reason. According to the solidity
 // spec https://solidity.readthedocs.io/en/latest/control-structures.html#revert,
-// the provided revert reason is abi-encoded as if it were a call to a function
-// `Error(string)`. So it's a special tool for it.
+// the provided revert reason is abi-encoded as if it were a call to function
+// `Error(string)` and `Panic(uint256)`. So it's a special tool for it.
```
Comment (rjl493456442) [inline accounts/abi/abi.go:274]: `Error(string)` or `Panic(uint256)`

File: accounts/abi/abi.go [hunk 3]
```diff
@@ -246,24 +247,65 @@ func (abi *ABI) HasReceive() bool {
...
+
+// panicReasons map is for readable panic codes
+// see this linkage for the deails
+// https://docs.soliditylang.org/en/v0.8.21/control-structures.html#panic-via-assert-and-error-via-require
+// the reason string list is copied from ether.js
+// https://github.com/ethers-io/ethers.js/blob/fa3a883ff7c88611ce766f58bdd4b8ac90814470/src.ts/abi/interface.ts#L207-L218
+var panicReasons = map[uint64]string{
+	0x00: "generic panic",
+	0x01: "assert(false)",
+	0x11: "arithmetic overflow",
```
Comment (rjl493456442) [inline accounts/abi/abi.go:261]: 0x11: If an arithmetic operation results in underflow or overflow outside of an unchecked { ... } block.

perhaps `0x11: "arithmetic underflow or overflow",`?

**Pull request merged**